src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Tuples.cs

```diff
@@ -674,5 +674,76 @@ public void TupleAttributeMissing()
+public struct S
+{
+}";
+            var comp = CreateCompilationWithMscorlib(text, references: s_attributeRefs);
+            comp.VerifyDiagnostics(
+                // (31,2): error CS8208: Cannot reference 'System.Runtime.CompilerServices.TupleElementNamesAttribute' explicitly. Use the tuple syntax to define tuple names.
+                // [TupleElementNames(new[] { "a", "b" })]
+                Diagnostic(ErrorCode.ERR_ExplicitTupleElementNames, @"TupleElementNames(new[] { ""a"", ""b"" })").WithLocation(31, 2),
+                // (5,2): error CS8208: Cannot reference 'System.Runtime.CompilerServices.TupleElementNamesAttribute' explicitly. Use the tuple syntax to define tuple names.
+                // [TupleElementNames(new[] { "a", "b" })]
+                Diagnostic(ErrorCode.ERR_ExplicitTupleElementNames, @"TupleElementNames(new[] { ""a"", ""b"" })").WithLocation(5, 2),
+                // (18,10): error CS8208: Cannot reference 'System.Runtime.CompilerServices.TupleElementNamesAttribute' explicitly. Use the tuple syntax to define tuple names.
+                //         [TupleElementNames(new[] { "x" })]ValueTuple<T> args);
+                Diagnostic(ErrorCode.ERR_ExplicitTupleElementNames, @"TupleElementNames(new[] { ""x"" })").WithLocation(18, 10),
+                // (11,6): error CS8208: Cannot reference 'System.Runtime.CompilerServices.TupleElementNamesAttribute' explicitly. Use the tuple syntax to define tuple names.
+                //     [TupleElementNames(new[] { "x", "y" })]
+                Diagnostic(ErrorCode.ERR_ExplicitTupleElementNames, @"TupleElementNames(new[] { ""x"", ""y"" })").WithLocation(11, 6),
+                // (14,14): error CS8208: Cannot reference 'System.Runtime.CompilerServices.TupleElementNamesAttribute' explicitly. Use the tuple syntax to define tuple names.
+                //     [return: TupleElementNames(new string[] { null, null })]
+                Diagnostic(ErrorCode.ERR_ExplicitTupleElementNames, "TupleElementNames(new string[] { null, null })").WithLocation(14, 14),
+                // (15,36): error CS8208: Cannot reference 'System.Runtime.CompilerServices.TupleElementNamesAttribute' explicitly. Use the tuple syntax to define tuple names.
+                //     public ValueTuple<int, int> M([TupleElementNames(new string[] { null})] ValueTuple x) => (0, 0);
+                Diagnostic(ErrorCode.ERR_ExplicitTupleElementNames, "TupleElementNames(new string[] { null})").WithLocation(15, 36),
+                // (20,6): error CS0592: Attribute 'TupleElementNames' is not valid on this declaration type. It is only valid on 'class, struct, property, indexer, field, parameter, return' declarations.
+                //     [TupleElementNames(new[] { "y" })]
+                Diagnostic(ErrorCode.ERR_AttributeOnBadSymbolType, "TupleElementNames").WithArguments("TupleElementNames", "class, struct, property, indexer, field, parameter, return").WithLocation(20, 6),
```

Comment:
I'm looking at the list of possible targets for this attribute:
`[AttributeUsage(AttributeTargets.Field | AttributeTargets.Parameter | AttributeTargets.Property | AttributeTargets.ReturnValue | AttributeTargets.Class | AttributeTargets.Struct )]` and matching that to the list of source symbols where you added a check.

One difference is that you added a check on events, but the attribute is currently not allowed on events.
This is visible in this error (lines 735 to 737).
Does this mean we should update the attribute (allow it to be placed on events)?

Btw, for reference, the list of all possible targets for an attribute: https://msdn.microsoft.com/en-us/library/system.attributetargets(v=vs.71).aspx